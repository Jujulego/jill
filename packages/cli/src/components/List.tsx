import { Box, Static, Text, useStdout } from 'ink';
import { ReactElement } from 'react';

// Types
export type ListItem<K extends string> = Partial<Record<K, string | number>> & Record<string, unknown>;

export interface ListProps<K extends string> {
	attrs: K[];
	data: ListItem<K>[];
	withoutHeaders?: boolean;
	json?: boolean;
}

// Utils
function capitalize(str: string): string {
	return `${str.at(0)?.toUpperCase()}${str.substring(1)}`;
}

// Component
export const List = <K extends string>(props: ListProps<K>): ReactElement => {
	const { attrs, data, withoutHeaders, json } = props;
	const { stdout } = useStdout();

	// Render
	if (json) {
		const lines = JSON.stringify(data, null, stdout?.isTTY ? 2 : 0).split('\n');

		return (
			<Static items={lines}>
				{ (line, idx) => <Text key={idx}>{ line }</Text> }
			</Static>
		);
	}

	return (
		<Box>
			{attrs.map((attr) => (
				<Box key={attr} flexDirection="column" marginRight={2}>
					{!withoutHeaders && <Text bold>{capitalize(attr)}</Text>}
					{data.map((d, idx) => (
						<Text key={idx} color={d[attr] ? '' : 'grey'}>{d[attr] || 'unset'}</Text>
					))}
				</Box>
			))}
		</Box>
	);
};
