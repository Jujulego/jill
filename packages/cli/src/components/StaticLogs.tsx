import { logger } from '@jujulego/jill-core';
import { FC, useEffect, useState } from 'react';
import { format } from 'winston';
import Transport from 'winston-transport';
import { Static, Text } from 'ink';

// Components
export const StaticLogs: FC = () => {
  // State
  const [logs, setLogs] = useState<any[]>([]);

  // Effects
  useEffect(() => {
    logger.add(new class extends Transport {
      // Constructor
      constructor() {
        super({
          format: format.combine(
            format.errors(),
            format.colorize({ message: true, colors: { debug: 'grey', verbose: 'blue', info: 'white', error: 'red' } }),
          )
        });
      }

      // Methods
      log(log: any, next: () => void): void {
        setLogs((old) => [...old, log]);
        next();
      }
    });
  }, []);

  // Render
  return (
    <Static items={logs}>
      { ({ label, message }, idx) => (
        <Text key={idx}>
          <Text color="grey">jill: { label && `[${label}] ` }</Text>
          { message }
        </Text>
      ) }
    </Static>
  );
};
