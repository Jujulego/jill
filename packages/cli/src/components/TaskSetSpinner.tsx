import { Task, TaskSet } from '@jujulego/jill-core';
import { TaskSpinner } from '@jujulego/jill-common';
import { FC, useEffect, useState } from 'react';

// Types
export interface TaskSetSpinnerProps {
  taskSet: TaskSet;
}

// Component
export const TaskSetSpinner: FC<TaskSetSpinnerProps> = (props) => {
  const { taskSet } = props;

  // State
  const [tasks, setTasks] = useState<Task[]>([]);

  // Effects
  useEffect(() => {
    const handler = (task: Task) => {
      setTasks((old) => [...old, task]);
    };

    taskSet.on('started', handler);

    return () => {
      taskSet.off('started', handler);
    };
  }, [taskSet]);

  // Render
  return (
    <>
      { tasks.map((task, idx) => <TaskSpinner key={idx} task={task} />) }
    </>
  );
};
