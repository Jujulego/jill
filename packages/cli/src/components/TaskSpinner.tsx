import { Task } from '@jujulego/jill-core';
import { Text } from 'ink';
import Spinner from 'ink-spinner';
import logSymbols from 'log-symbols';
import { FC, useEffect, useState } from 'react';

// Types
export interface TaskSpinnerProps {
  task: Task;
}

// Component
export const TaskSpinner: FC<TaskSpinnerProps> = (props) => {
  const { task } = props;

  // State
  const [status, setStatus] = useState(task.status);

  // Effects
  useEffect(() => {
    task.on('status', setStatus);

    return () => {
      task.off('status', setStatus);
    };
  }, [task]);

  // Render
  if (status === 'done') {
    return (
      <Text>
        <Text color="green">{ logSymbols.success }</Text>{' '}
        <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
        { task.name } is { status }
      </Text>
    );
  } else if (status === 'failed') {
    return (
      <Text>
        <Text color="red">{ logSymbols.error }</Text>{' '}
        <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
        { task.name } is { status }
      </Text>
    );
  }

  return (
    <Text>
      <Spinner type={status === 'running' ? 'line' : 'simpleDotsScrolling'} />{' '}
      <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
      { task.name } is { status }
    </Text>
  );
};
