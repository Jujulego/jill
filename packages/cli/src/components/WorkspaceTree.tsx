import { Workspace } from '@jujulego/jill-core';
import { Newline, Text, TextProps } from 'ink';
import { FC, ReactElement, useEffect, useState } from 'react';

// Types
export interface WorkspaceTreeProps {
  workspace: Workspace;
  dev?: boolean;
  level?: ReactElement;
}

// Utils
const style = (dev: boolean): TextProps => ({ color: dev ? 'blue' : '' });

// Component
export const WorkspaceTree: FC<WorkspaceTreeProps> = (props) => {
  const { workspace: wks, dev = false, level = '' } = props;

  // State
  const [deps, setDeps] = useState<[Workspace, boolean][]>([]);

  // Effects
  useEffect(() => void (async () => {
    const deps: [Workspace, boolean][] = [];

    for await (const dep of wks.dependencies()) {
      deps.push([dep, dev]);
    }

    for await (const dep of wks.devDependencies()) {
      deps.push([dep, true]);
    }

    setDeps(deps);
  })(), [wks]);

  // Render
  return (
    <Text>
      <Text {...style(dev)}>{ wks.name }</Text>
      { wks.version && (<Text color="grey">@{ wks.version }</Text>) }
      <Newline />

      { deps.map(([dep, isDev], idx) => (
        <Text key={dep.name}>
          { level }<Text {...style(dev)}>{ idx === deps.length - 1 ? '└' : '├'}─{' '}</Text>
          <WorkspaceTree
            workspace={dep}
            dev={isDev}
            level={<>{ level }<Text {...style(dev)}>{ idx === deps.length - 1 ? ' ' : '│' }{'  '}</Text></>}
          />
        </Text>
      )) }
    </Text>
  );
};
