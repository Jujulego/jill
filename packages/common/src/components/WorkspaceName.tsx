import { Workspace } from '@jujulego/jill-core';
import { Text, TextProps } from 'ink';
import { FC } from 'react';

// Types
export interface WorkspaceNameProps extends TextProps {
  workspace: Workspace;
}

// Component
export const WorkspaceName: FC<WorkspaceNameProps> = ({ workspace, ...style }) => (
  <Text {...style}>
    { workspace.name }
    { workspace.version && (<Text color="grey">@{ workspace.version }</Text>) }
  </Text>
);
