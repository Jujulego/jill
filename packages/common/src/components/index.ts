export * from './List';
export * from './TaskSpinner';
export * from './TaskSetSpinner';
