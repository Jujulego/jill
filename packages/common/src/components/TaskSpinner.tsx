import { Task } from '@jujulego/jill-core';
import { Text } from 'ink';
import Spinner from 'ink-spinner';
import logSymbols from 'log-symbols';
import { FC, useEffect, useState } from 'react';
import { WorkspaceName } from './WorkspaceName';

// Types
export interface TaskSpinnerProps {
  task: Task;
}

// Component
export const TaskSpinner: FC<TaskSpinnerProps> = (props) => {
  const { task } = props;
  const wks = task.context.workspace;

  // State
  const [status, setStatus] = useState(task.status);

  // Effects
  useEffect(() => {
    task.on('status', setStatus);

    return () => {
      task.off('status', setStatus);
    };
  }, [task]);

  // Render
  if (status === 'done') {
    return (
      <Text>
        { logSymbols.success }{' '}
        { wks && <><Text color="grey">[<WorkspaceName workspace={wks} />]</Text>{' '}</> }
        { task.name } is { status }
      </Text>
    );
  } else if (status === 'failed') {
    return (
      <Text>
        { logSymbols.error }{' '}
        { wks && <><Text color="grey">[<WorkspaceName workspace={wks} />]</Text>{' '}</> }
        { task.name } is { status }
      </Text>
    );
  }

  return (
    <Text>
      <Spinner type={status === 'running' ? 'line' : 'simpleDotsScrolling'} />{' '}
      { wks && <><Text color="grey">[<WorkspaceName workspace={wks} />]</Text>{' '}</> }
      { task.name } is { status }
    </Text>
  );
};
