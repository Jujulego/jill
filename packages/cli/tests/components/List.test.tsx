import chalk from 'chalk';
import { render } from 'ink-testing-library';

import { List } from '../../src/components/List';

// Constants
const TEST_DATA = [
  { id: 85, name: 'Test n°85', result: 'success' },
  { id: 128, name: 'Test n°128', result: 'failed' },
  { id: 223, name: 'Test n°223' },
];

// Setup
chalk.level = 1;

// Tests
describe('List', () => {
  it('should render data in aligned column', () => {
    // Render
    const { lastFrame } = render(
      <List
        attrs={['id', 'result']}
        data={TEST_DATA}
      />
    );

    // Checks
    expect(lastFrame()).toBe(
      chalk`{bold Id}   {bold Result}\n` +
      chalk`85   success\n` +
      chalk`128  failed\n` +
      chalk`223  {grey unset}`
    );
  });

  it('should render without headers', () => {
    // Render
    const { lastFrame } = render(
      <List
        attrs={['id', 'result']}
        data={TEST_DATA}
        withoutHeaders
      />
    );

    // Checks
    expect(lastFrame()).toBe(
      chalk`85   success\n` +
      chalk`128  failed\n` +
      chalk`223  {grey unset}`
    );
  });

  it('should render as json', async () => {
    jest.spyOn(process.stdout, 'write');

    // Render
    const { lastFrame } = render(
      <List
        attrs={['id', 'result']}
        data={TEST_DATA}
        json
      />
    );
    await new Promise(res => setTimeout(res, 0));

    // Checks
    expect(lastFrame()).toMatchSnapshot();
  });
});
