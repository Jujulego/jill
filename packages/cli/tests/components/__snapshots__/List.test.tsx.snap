// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render as indented json 1`] = `
"[
  {
    \\"id\\": 85,
    \\"name\\": \\"Test nÂ°85\\",
    \\"result\\": \\"success\\"
  },
  {
    \\"id\\": 128,
    \\"name\\": \\"Test nÂ°128\\",
    \\"result\\": \\"failed\\"
  },
  {
    \\"id\\": 223,
    \\"name\\": \\"Test nÂ°128\\"
  }
]
"
`;

exports[`List should render as json 1`] = `
"[{\\"id\\":85,\\"name\\":\\"Test nÂ°85\\",\\"result\\":\\"success\\"},{\\"id\\":128,\\"name\\":\\"Test
nÂ°128\\",\\"result\\":\\"failed\\"},{\\"id\\":223,\\"name\\":\\"Test nÂ°128\\"}]
"
`;

exports[`List should render data in aligned column 1`] = `
"[1mId[22m   [1mResult[22m
85   success
128  failed
223  [90munset[39m"
`;

exports[`List should render without headers 1`] = `
"85   success
128  failed
223  [90munset[39m"
`;
